third_party/move/move-bytecode-verifier/src/type_safety.rs

```diff
@@ -300,6 +301,115 @@ fn call(
     Ok(())
 }
 
+fn clos_eval(
+    verifier: &mut TypeSafetyChecker,
+    meter: &mut impl Meter,
+    offset: CodeOffset,
+    expected_ty: &SignatureToken,
+) -> PartialVMResult<()> {
+    let SignatureToken::Function(param_tys, ret_tys, abilities) = expected_ty else {
+        // The signature checker has ensured this is a function
+        safe_assert!(false);
+        unreachable!()
+    };
+    // On top of the stack is the closure, pop it.
+    let closure_ty = safe_unwrap!(verifier.stack.pop());
+    // Verify that the closure type matches the expected type
+    if &closure_ty != expected_ty {
+        return Err(verifier
+            .error(StatusCode::CALL_TYPE_MISMATCH_ERROR, offset)
+            .with_message("closure type mismatch".to_owned()));
+    }
+    // Verify that the abilities match
+    let SignatureToken::Function(_, _, closure_abilities) = closure_ty else {
```

Comment:
Overall, this code is quite confusing, I would prefer it written differently if possible. For example:

```
let closure_ty = safe_unwrap!(verifier.stack.pop());
match (closure_ty, actual_ty) {
  (Function(..), Function(..)) => {
      // do all checks here
  },
  _ => return Err(...)
}
```